### PR TITLE
block_hotplug_check:Fix unknown version of qemu issue

### PR DESCRIPTION
**Pull request by qingwangrh**
ID:1733761,1511800
1.Add blockdev support
2.Improve test efficiency
3.Remove redunant command check
4.Add q35 support
Signed-off-by: qingwangrh <qinwang@redhat.com>

### CODE REVIEW

**Comment by qingwangrh** [comment]
python ConfigTest.py --testcase=block_hotplug_check --guestname=Win2019,RHEL.8.4.0 --driveformat=virtio_scsi,virtio_blk  --clone=no
 (1/4) type_specific.io-github-autotest-qemu.block_hotplug_check: PASS (227.41 s)
 (2/4) type_specific.io-github-autotest-qemu.block_hotplug_check: PASS (484.06 s)
 (3/4) type_specific.io-github-autotest-qemu.block_hotplug_check: PASS (229.79 s)
 (4/4) type_specific.io-github-autotest-qemu.block_hotplug_check: PASS (481.81 s)

**Comment by qingwangrh** [comment]
@YongxueHong @zhenyzha @menli820 please help to review, thanks.

**Comment by PaulYuuu** [comment]
Cc @maxujun

**Comment by zhenyzha** [comment]
(1/2) type_specific.io-github-autotest-qemu.block_hotplug_check: ERROR: Timeout expired while waiting for shell command to complete: 'cat /proc/partitions |tee /tmp/partitions'    (output: '') (389.59 s)
 (2/2) type_specific.io-github-autotest-qemu.block_hotplug_check: PASS (430.65 s)

Need to confirm, it may be a product problem. Paste the detailed log into the autobug

**Comment by menli820** [comment]
hit issue on q35 as confirmed with you.
[root@hp-dl385g10-18 kar]#  python3  ConfigTest.py --testcase=block_hotplug_check --guestname=Win2016 --platform=x86_64  --machines=q35
 (1/1) type_specific.io-github-autotest-qemu.block_hotplug_check: ERROR: QMP command 'device_add' failed    (arguments: {'id': 'block-idigxLoU', 'driver': 'virtio-blk-pci', 'drive': 'block-idigxLoU'},    error message: {'class': 'GenericError', 'desc': "Bus 'pcie.0' does not support hotplugging"}) (84.26 s)

good for me on pc, thanks

**Comment by maxujun** [comment]
Hit this situation
 (04/10) repeat4.Host_RHEL.m8.u4.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.ppc64le.io-github-autotest-qemu.block_hotplug_check: ERROR: Shell process terminated while waiting for command to complete: 'cat /proc/partitions |tee /tmp/partitions'    (status: 255,    output: 'client_loop: send disconnect: Broken pipe\n') (359.53 s)
log
06:22:53 ERROR|     raise ShellProcessTerminatedError(cmd, error.status, output)
06:22:53 ERROR| aexpect.exceptions.ShellProcessTerminatedError: Shell process terminated while waiting for command to complete: 'cat /proc/partitions |tee /tmp/partitions'    (status: 255,    output: 'client_loop: send disconnect: Broken pipe\n')

Just like the same situation with zhenyzha's

**Comment by qingwangrh** [comment]
python ConfigTest.py --testcase=block_hotplug_check --guestname=Win2019,RHEL.8.4.0 --driveformat=virtio_blk,virtio_scsi --machines=q35,i440fx --clone=no
(1/4) type_specific.io-github-autotest-qemu.block_hotplug_check: PASS (438.71 s)
 (2/4) type_specific.io-github-autotest-qemu.block_hotplug_check: PASS (530.12 s)
 (3/4) type_specific.io-github-autotest-qemu.block_hotplug_check: PASS (679.77 s)
 (4/4) type_specific.io-github-autotest-qemu.block_hotplug_check: PASS (943.70 s)


Patch updated ,Please help to review @menli820 @zhenyzha @maxujun

**Comment by qingwangrh** [comment]
> Hit this situation
> (04/10) repeat4.Host_RHEL.m8.u4.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.ppc64le.io-github-autotest-qemu.block_hotplug_check: ERROR: Shell process terminated while waiting for command to complete: 'cat /proc/partitions |tee /tmp/partitions' (status: 255, output: 'client_loop: send disconnect: Broken pipe\n') (359.53 s)
> log
> 06:22:53 ERROR| raise ShellProcessTerminatedError(cmd, error.status, output)
> 06:22:53 ERROR| aexpect.exceptions.ShellProcessTerminatedError: Shell process terminated while waiting for command to complete: 'cat /proc/partitions |tee /tmp/partitions' (status: 255, output: 'client_loop: send disconnect: Broken pipe\n')
> 
> Just like the same situation with zhenyzha's

It looks like product bug on ppc Bug 1902548 - When there are more than 7 virtio-blk devices hot-inserted, HMP reports an error: virtio-blk failed to set guest notifier (-24), ensure -accel kvm is set.

**Comment by zhenyzha** [comment]
python3 ConfigTest.py --testcase=block_hotplug_check --guestname=RHEL.8.4.0 --driveformat=virtio_scsi,virtio_blk --clone=no

 (1/1) type_specific.io-github-autotest-qemu.block_hotplug_check: PASS (556.01 s)

LGTM.Acked-by: Zhenyu Zhang <zhenyzha@redhat.com>

**Comment by maxujun** [comment]
Hi @qingwangrh Why only virtio_blk?I think we shouldn't limit block device type in this case.

**Comment by maxujun** [comment]
> 
> 
> > Hit this situation
> > (04/10) repeat4.Host_RHEL.m8.u4.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.ppc64le.io-github-autotest-qemu.block_hotplug_check: ERROR: Shell process terminated while waiting for command to complete: 'cat /proc/partitions |tee /tmp/partitions' (status: 255, output: 'client_loop: send disconnect: Broken pipe\n') (359.53 s)
> > log
> > 06:22:53 ERROR| raise ShellProcessTerminatedError(cmd, error.status, output)
> > 06:22:53 ERROR| aexpect.exceptions.ShellProcessTerminatedError: Shell process terminated while waiting for command to complete: 'cat /proc/partitions |tee /tmp/partitions' (status: 255, output: 'client_loop: send disconnect: Broken pipe\n')
> > Just like the same situation with zhenyzha's
> 
> It looks like product bug on ppc Bug 1902548 - When there are more than 7 virtio-blk devices hot-inserted, HMP reports an error: virtio-blk failed to set guest notifier (-24), ensure -accel kvm is set.

Hi ,I retested it,still failed but no the problem you said.so I thinks there must be something wrong with 'cat /proc/partitions |tee /tmp/partitions'.

**Comment by qingwangrh** [comment]
> Hi @qingwangrh Why only virtio_blk?I think we shouldn't limit block device type in this case.

from original design it used only virtio_blk filter and code only support blk disk hotplug.

**Comment by qingwangrh** [comment]
> > > Hit this situation
> > > (04/10) repeat4.Host_RHEL.m8.u4.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.ppc64le.io-github-autotest-qemu.block_hotplug_check: ERROR: Shell process terminated while waiting for command to complete: 'cat /proc/partitions |tee /tmp/partitions' (status: 255, output: 'client_loop: send disconnect: Broken pipe\n') (359.53 s)
> > > log
> > > 06:22:53 ERROR| raise ShellProcessTerminatedError(cmd, error.status, output)
> > > 06:22:53 ERROR| aexpect.exceptions.ShellProcessTerminatedError: Shell process terminated while waiting for command to complete: 'cat /proc/partitions |tee /tmp/partitions' (status: 255, output: 'client_loop: send disconnect: Broken pipe\n')
> > > Just like the same situation with zhenyzha's
> > 
> > 
> > It looks like product bug on ppc Bug 1902548 - When there are more than 7 virtio-blk devices hot-inserted, HMP reports an error: virtio-blk failed to set guest notifier (-24), ensure -accel kvm is set.
> 
> Hi ,I retested it,still failed but no the problem you said.so I thinks there must be something wrong with 'cat /proc/partitions |tee /tmp/partitions'.

I think the guest became hang status or crash result in command execution failed. 
You may try it manually just plug one disk then execute above command it should work.

**Comment by menli820** [comment]
LGTM.

(1/1) Host_RHEL.m8.u3.product_av.qcow2.virtio_blk.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.block_hotplug_check: PASS (676.13 s)
(2/2) Host_RHEL.m8.u3.product_av.qcow2.virtio_blk.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.block_hotplug_check.q35: PASS (941.93 s)

**Comment by maxujun** [comment]
LGTM.
1/1) Host_RHEL.m8.u4.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.4.0.ppc64le.io-github-autotest-qemu.block_hotplug_check: PASS (563.57 s)

**Comment by YongxueHong** [comment]
Suggest your commit header: `block_hotplug_check: Fix unknown version of qemu issue`

File: qemu/tests/pci_hotplug_check.py [hunk 1]
```diff
@@ -139,15 +146,24 @@ def device_add_block(pci_num, queues=1, pci_id=None):
...
 
-        verify_supported_device(pci_model)
-        if drive_cmd_type == "drive_add":
+        if nonlocal_vars["verify_device_flag"]:
+            verify_supported_device(pci_model)
+        nonlocal_vars["verify_device_flag"] = False
+
+        if drive_cmd_type == "blockdev-add":
+            driver_add_cmd = (
+                "{0} driver=file,filename={1},node-name=file_{2};{0} driver={3},node-name={2},file=file_{2}".format(
```
[the remark this answers is not in the document]
Comment (qingwangrh) [reply qemu/tests/pci_hotplug_check.py]: updated

**Review by YongxueHong — APPROVED** [review]
LGTM.

**Pull request merged**